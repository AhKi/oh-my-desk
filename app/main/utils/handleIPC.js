import { ipcMain } from 'electron';
import { openPreference } from 'main/utils/window/preference';

function handleIPC() {
  ipcMain.on('preference.open', () => {
    openPreference();
  });
}

export default handleIPC;
