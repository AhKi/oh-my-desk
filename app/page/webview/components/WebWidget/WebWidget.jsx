import React from 'react';
import PropTypes from 'prop-types';
import * as USER_AGENT from 'constants/userAgent';
import WebWidgetHeader from './components/WebWidgetHeader';
import WebWidgetMobileHeader from './components/WebWidgetMobileHeader';
import WebWidgetSetting from './components/WebWidgetSetting';
import './WebWidget.scss';

const propTypes = {
  widget: PropTypes.shape({
    name: PropTypes.string,
    url: PropTypes.string,
  }),
  onOpenPreference: PropTypes.func,
  onUpdateInfo: PropTypes.func, // eslint-disable-line
};
const defaultProps = {
  widget: {},
  onOpenPreference() {},
  onUpdateInfo() {},
};

class WebWidget extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: false,
      isSettingOpen: false,
      isMobileHeaderOpen: true,
    };
    this.webViewRef = React.createRef();
    this.prevScrollY = 0;
    this.tick = null;
    this.setKeyEvent = this.setKeyEvent.bind(this);
    this.toggleIsOnTop = this.toggleIsOnTop.bind(this);
    this.handleWidgetGoBack = this.handleWidgetGoBack.bind(this);
    this.handleWidgetGoForward = this.handleWidgetGoForward.bind(this);
    this.handleWidgetRefresh = this.handleWidgetRefresh.bind(this);
    this.handleWidgetStopRefresh = this.handleWidgetStopRefresh.bind(this);
    this.handleToggleSettingMenu = this.handleToggleSettingMenu.bind(this);
  }

  componentDidMount() {
    document.addEventListener('keydown', this.setKeyEvent);
    // add event when webview page loading
    this.webViewRef.current.addEventListener('did-start-loading', () => {
      this.setState({ isLoading: true });
    });
    this.webViewRef.current.addEventListener('did-finish-load', () => {
      this.setState({ isLoading: false });
    });
    this.webViewRef.current.addEventListener('did-stop-loading', () => {
      this.setState({ isLoading: false });
    });
    // Communicate webview tag and BrowserWindow
    // TODO fix communication logic
    this.webViewRef.current.addEventListener('console-message', (res) => {
      const scrollY = Number(res.message);
      if (typeof scrollY !== 'number') {
        return;
      }

      if (scrollY === 0 || this.prevScrollY === 0) {
        clearTimeout(this.tick);
        this.tick = null;
        this.setState({ isMobileHeaderOpen: true });
      } else if (Math.abs(scrollY - this.prevScrollY) > 30) {
        clearTimeout(this.tick);
        this.setState({ isMobileHeaderOpen: true });
        this.tick = setTimeout(() => this.setState({ isMobileHeaderOpen: false }), 2000);
      } else if (!this.tick) {
        this.setState({ isMobileHeaderOpen: false });
      }
      this.prevScrollY = scrollY;
    });
    this.webViewRef.current.addEventListener('dom-ready', () => {
      this.webViewRef.current.executeJavaScript(`
        window.addEventListener('scroll', () => {
          console.log(window.scrollY);
        });
      `, true);
    });
  }

  componentDidUpdate(prevProps) {
    const { widget } = this.props;
    const userAgent = widget.isMobile ? USER_AGENT.MOBILE : USER_AGENT.DESKTOP;

    if (prevProps.widget.url !== widget.url) {
      this.webViewRef.current.loadURL(widget.url, { userAgent });
    }

    if (prevProps.widget.isMobile !== widget.isMobile) {
      this.webViewRef.current.setUserAgent(userAgent);
    }
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.setKeyEvent);
  }

  setKeyEvent(e) {
    if ((e.metaKey || e.altKey) && e.keyCode === 37) {
      // cmd + arrowLeft (mac)
      // alt + arrowLeft (window)
      this.webViewRef.current.goBack();
    } else if ((e.metaKey || e.altKey) && e.keyCode === 39) {
      // cmd + arrowRight (mac)
      // alt + arrowRight (window)
      this.webViewRef.current.goForward();
    } else if ((e.metaKey || e.altKey) && e.keyCode === 82) {
      // cmd + R (mac)
      // alt + R (window)
      this.webViewRef.current.reload();
    }
  }

  toggleIsOnTop() {
    const { widget, onUpdateInfo } = this.props;

    onUpdateInfo(widget.id, { isOnTop: !widget.isOnTop });
  }

  handleWidgetGoBack() {
    this.webViewRef.current.goBack();
  }

  handleWidgetGoForward() {
    this.webViewRef.current.goForward();
  }

  handleWidgetRefresh() {
    this.webViewRef.current.reload();
  }

  handleWidgetStopRefresh() {
    this.webViewRef.current.stop();
  }

  handleToggleSettingMenu(bool) {
    if (typeof bool === 'boolean') {
      this.setState({ isSettingOpen: bool });
    } else {
      this.setState(nextState => ({ isSettingOpen: !nextState.isSettingOpen }));
    }
  }

  render() {
    const {
      isLoading,
      isSettingOpen,
      isMobileHeaderOpen,
    } = this.state;
    const {
      widget,
      onUpdateInfo,
      onOpenPreference,
    } = this.props;

    return (
      <div className="WebWidget">
        <WebWidgetHeader
          webView={this.webViewRef.current}
          title={widget.name}
          isLoading={isLoading}
          isOnTop={widget.isOnTop}
          onToggleIsOnTop={this.toggleIsOnTop}
          onGoBack={this.handleWidgetGoBack}
          onGoForward={this.handleWidgetGoForward}
          onRefresh={this.handleWidgetRefresh}
          onStopRefresh={this.handleWidgetStopRefresh}
          onToggleSetting={this.handleToggleSettingMenu}
        />
        <WebWidgetMobileHeader
          webView={this.webViewRef.current}
          isLoading={isLoading}
          isOpen={isMobileHeaderOpen || isSettingOpen}
          onGoBack={this.handleWidgetGoBack}
          onGoForward={this.handleWidgetGoForward}
          onRefresh={this.handleWidgetRefresh}
          onStopRefresh={this.handleWidgetStopRefresh}
          onToggleSetting={this.handleToggleSettingMenu}
        />
        {isSettingOpen && (
          <WebWidgetSetting
            name={widget.name}
            widget={widget}
            url={widget.url}
            onToggleSetting={this.handleToggleSettingMenu}
            onUpdateInfo={onUpdateInfo}
            onOpenPreference={onOpenPreference}
          />
        )}
        <webview
          ref={this.webViewRef}
          src="https://www.github.com"
          style={{
            display: 'inline-flex',
            width: '100%',
            height: '100%',
            overflow: 'auto',
            overflowY: 'hidden',
          }}
        />
      </div>
    );
  }
}

WebWidget.propTypes = propTypes;
WebWidget.defaultProps = defaultProps;

export default WebWidget;
