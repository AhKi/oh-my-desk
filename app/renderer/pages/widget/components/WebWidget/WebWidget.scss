.WebWidget {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;
}

.WebWidget__content {
  flex: 1;
  display: flex;
}

.WebWidget__webview {
  flex: 1;
}
