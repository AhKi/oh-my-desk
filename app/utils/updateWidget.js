import * as IPC from 'constants/ipc';

function updateWidget(type, info) {
  window.ipcRenderer.send(IPC.WIDGET_MANAGE, {
    operation: 'UPDATE',
    widget: {
      id: info.id,
      type,
      name: info.name,
      url: info.url,
      position: info.position && {
        x: info.position.x,
        y: info.position.y,
      },
      size: info.size && {
        width: info.size.width,
        height: info.size.height,
      },
      transparency: info.transparency,
      isActive: info.isActive,
      isIcon: info.isIcon,
      isOnTop: info.isOnTop,
      favicon: info.favicon,
    },
  });
  window.ipcRenderer.send(IPC.WIDGET_INFO_REQUEST);
}

export default updateWidget;
