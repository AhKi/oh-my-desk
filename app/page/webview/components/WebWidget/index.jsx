import React from 'react';
import PropTypes from 'prop-types';
import url from 'url';
import * as PATH from 'constants/path';
import * as USER_AGENT from 'constants/userAgent';
import ModalContainer from 'page/Components/Modal/ModalContainer';
import widgetContextMenu from 'utils/process/widgetContextMenu';
import WidgetHeaderContainer from '../../containers/WidgetHeaderContainer';
import MenuNewWindow from '../MenuNewWindow';
import './WebWidget.scss';

const propTypes = {
  defaultUserAgent: PropTypes.string,
  widget: PropTypes.shape({
    name: PropTypes.string,
    url: PropTypes.string,
  }),
};
const defaultProps = {
  defaultUserAgent: 'DESKTOP',
  widget: {},
};

class WebWidget extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentUrl: '',
      isLoading: false,
      isSettingOpen: false,
      newWindowURL: null,
    };
    this.webViewRef = React.createRef();
    this.prevScrollY = 0;
    this.tick = null;
    this.mousePosition = {
      x: 0,
      y: 0,
    };
    this.handleToggleSettingMenu = this.handleToggleSettingMenu.bind(this);
    this.handleToggleNewWindowMenu = this.handleToggleNewWindowMenu.bind(this);
  }

  componentDidMount() {
    const webView = this.webViewRef.current;
    // add event when webview page loading
    webView.addEventListener('did-start-loading', () => {
      this.setState({ isLoading: true });
    });
    webView.addEventListener('did-finish-load', () => {
      this.setState({ isLoading: false });
    });
    webView.addEventListener('did-stop-loading', () => {
      this.setState({ isLoading: false });
    });
    webView.addEventListener('new-window', (e) => {
      this.handleToggleNewWindowMenu(e.url);
    });
    webView.addEventListener('did-navigate', (e) => {
      this.setState({ currentUrl: e.url });
    });
    webView.addEventListener('did-navigate-in-page', (e) => {
      this.setState({ currentUrl: e.url });
    });
    window.addEventListener('contextmenu', () => {
      widgetContextMenu(this.webViewRef.current);
    });

    // webView.addEventListener('dom-ready', () => {
    //   this.webViewRef.current.openDevTools();
    // });

    window.addEventListener('mousemove', (e) => {
      this.mousePosition = {
        x: e.pageX,
        y: e.pageY,
      };
    });
  }

  componentDidUpdate(prevProps) {
    const { currentUrl } = this.state;
    const { defaultUserAgent, widget } = this.props;
    let userAgent;
    if (widget.userAgent) {
      userAgent = widget.userAgent === 'MOBILE' ? USER_AGENT.MOBILE : USER_AGENT.DESKTOP;
    } else {
      userAgent = defaultUserAgent === 'MOBILE' ? USER_AGENT.MOBILE : USER_AGENT.DESKTOP;
    }

    if (prevProps.widget.url !== widget.url) {
      this.webViewRef.current.loadURL(widget.url, { userAgent });
    }

    if (
      (!widget.userAgent && prevProps.defaultUserAgent !== defaultUserAgent) ||
      (prevProps.widget.userAgent !== widget.userAgent)
    ) {
      // Most website provide separate url about userAgent.
      // Mobile page url is prepended with 'm.'.
      // So remove 'm.' in url when change userAgent mobile to desktop.
      this.webViewRef.current.loadURL(
        userAgent === USER_AGENT.DESKTOP ? currentUrl.replace('https://m.', 'https://') : currentUrl,
        { userAgent },
      );
    }
  }

  handleToggleSettingMenu(bool) {
    if (typeof bool === 'boolean') {
      this.setState({ isSettingOpen: bool });
    } else {
      this.setState(prevState => ({ isSettingOpen: !prevState.isSettingOpen }));
    }
  }

  handleToggleNewWindowMenu(targetURL) {
    if (targetURL) {
      this.setState({ newWindowURL: targetURL });
    } else {
      this.setState({ newWindowURL: '' });
    }
  }

  render() {
    const {
      currentUrl,
      isLoading,
      newWindowURL,
    } = this.state;
    const {
      defaultUserAgent,
      widget,
    } = this.props;

    return (
      <div className="WebWidget">
        <ModalContainer />
        <WidgetHeaderContainer
          currentUrl={currentUrl}
          defaultUserAgent={defaultUserAgent}
          webView={this.webViewRef.current}
          title={widget.name}
          url={widget.url}
          id={widget.id}
          isLoading={isLoading}
          isOnTop={widget.isOnTop}
          userAgent={widget.userAgent}
          onToggleSetting={this.handleToggleSettingMenu}
        />
        {newWindowURL && (
          <MenuNewWindow
            url={newWindowURL}
            webview={this.webViewRef.current}
            x={this.mousePosition.x}
            y={this.mousePosition.y}
            onClose={this.handleToggleNewWindowMenu}
          />
        )}
        <webview
          ref={this.webViewRef}
          src="https://www.github.com"
          style={{
            display: 'inline-flex',
            width: '100%',
            height: '100%',
            overflow: 'auto',
            overflowY: 'hidden',
          }}
          preload={url.format({
            pathname: PATH.PRELOAD_SCRIPT_PATH,
            protocol: 'file:',
            slashed: true,
          })}
        />
      </div>
    );
  }
}

WebWidget.propTypes = propTypes;
WebWidget.defaultProps = defaultProps;

export default WebWidget;
