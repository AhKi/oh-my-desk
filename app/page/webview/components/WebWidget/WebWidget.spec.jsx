import React from 'react';
import { mount } from 'enzyme';

import * as utils from 'utils/updateWidget';
import WebWidget from '.';

describe('<WebWidget />', () => {
  const webView = {
    canGoBack() {},
    canGoForward() {},
    goBack() {},
    goForward() {},
    reload() {},
    stop() {},
    isMaximized() {},
  };
  const ipcRenderer = {
    on: () => {},
    send: () => {},
  };
  const remote = {
    getCurrentWindow: () => {},
  };
  window.ipcRenderer = ipcRenderer;
  window.remote = remote;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match to snapshot when render default', () => {
    const wrapper = mount(<WebWidget />);
    wrapper.instance().webViewRef = webView;

    expect(wrapper).toMatchSnapshot();
  });

  it('should match to snapshot when state.isSettingOpen === true', () => {
    const wrapper = mount(<WebWidget />);
    wrapper.instance().webViewRef = webView;
    wrapper.setState({
      isMobileHeaderOpen: false,
      isSettingOpen: true,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('test componentWillMount', () => {
    const componentWillMount = jest.spyOn(WebWidget.prototype, 'componentWillMount');
    const on = jest.spyOn(ipcRenderer, 'on');
    const wrapper = mount(<WebWidget />);
    wrapper.instance().webViewRef = webView;


    expect(componentWillMount).toHaveBeenCalledTimes(1);
    expect(on).toHaveBeenCalledTimes(1);
    expect(on).toHaveBeenCalledWith('widget-info', expect.any(Function));
  });

  it('test componnetDidMount and componentUnmount', () => {
    const componentDidMount = jest.spyOn(WebWidget.prototype, 'componentDidMount');
    const componentWillUnmount = jest.spyOn(WebWidget.prototype, 'componentWillUnmount');
    global.document.addEventListener = jest.fn();
    global.document.removeEventListener = jest.fn();
    const wrapper = mount(<WebWidget />);
    wrapper.instance().webViewRef = webView;
    const { setKeyEvent } = wrapper.instance();

    expect(componentDidMount).toHaveBeenCalledTimes(1);
    expect(global.document.addEventListener).toHaveBeenCalledTimes(1);
    expect(global.document.addEventListener).toHaveBeenCalledWith(
      'keydown',
      setKeyEvent,
    );

    wrapper.unmount();
    expect(componentWillUnmount).toHaveBeenCalledTimes(1);
    expect(global.document.removeEventListener).toHaveBeenCalledTimes(1);
    expect(global.document.removeEventListener).toHaveBeenCalledWith(
      'keydown',
      setKeyEvent,
    );
  });

  describe('should test setKeyEvent', () => {
    const goBack = jest.spyOn(webView, 'goBack');
    const goForward = jest.spyOn(webView, 'goForward');
    const reload = jest.spyOn(webView, 'reload');
    const wrapper = mount(<WebWidget />);
    wrapper.instance().webViewRef = webView;

    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('when alt + arrowLeft', () => {
      wrapper.instance().setKeyEvent({
        metaKey: true,
        altKey: true,
        keyCode: 37,
      });

      expect(goBack).toHaveBeenCalledTimes(1);
      expect(goBack).toHaveBeenCalledWith();

      wrapper.instance().setKeyEvent({
        altKey: true,
        keyCode: 37,
      });

      expect(goBack).toHaveBeenCalledTimes(2);
      expect(goBack).toHaveBeenCalledWith();
    });

    it('when alt + arrowRight', () => {
      wrapper.instance().setKeyEvent({
        metaKey: true,
        altKey: true,
        keyCode: 39,
      });

      expect(goForward).toHaveBeenCalledTimes(1);
      expect(goForward).toHaveBeenCalledWith();

      wrapper.instance().setKeyEvent({
        altKey: true,
        keyCode: 39,
      });

      expect(goForward).toHaveBeenCalledTimes(2);
      expect(goForward).toHaveBeenCalledWith();
    });

    it('when alt + R', () => {
      wrapper.instance().setKeyEvent({
        metaKey: true,
        altKey: true,
        keyCode: 82,
      });

      expect(reload).toHaveBeenCalledTimes(1);
      expect(reload).toHaveBeenCalledWith();

      wrapper.instance().setKeyEvent({
        altKey: true,
        keyCode: 82,
      });

      expect(reload).toHaveBeenCalledTimes(2);
      expect(reload).toHaveBeenCalledWith();
    });

    it('default', () => {
      wrapper.instance().setKeyEvent({
        altKey: false,
        keyCode: 82,
      });

      expect(goBack).toHaveBeenCalledTimes(0);
      expect(goForward).toHaveBeenCalledTimes(0);
      expect(reload).toHaveBeenCalledTimes(0);
    });
  });

  it('should test toggleIsOnTop', () => {
    const updateWidget = jest.spyOn(utils, 'default');
    const wrapper = mount(<WebWidget />);
    wrapper.instance().webViewRef = webView;

    wrapper.instance().toggleIsOnTop();

    expect(updateWidget).toHaveBeenCalledTimes(1);
    expect(updateWidget).toHaveBeenCalledWith(
      'web',
      {
        isOnTop: true,
      },
    );
  });

  describe('test WebWidget handle function', () => {
    const goBack = jest.spyOn(webView, 'goBack');
    const goForward = jest.spyOn(webView, 'goForward');
    const reload = jest.spyOn(webView, 'reload');
    const stop = jest.spyOn(webView, 'stop');
    const wrapper = mount(<WebWidget />);
    wrapper.instance().webViewRef = webView;

    it('when call handleWidgetGoBack', () => {
      wrapper.instance().handleWidgetGoBack();

      expect(goBack).toHaveBeenCalledTimes(1);
      expect(goBack).toHaveBeenCalledWith();
    });

    it('when call handleWidgetGoFront', () => {
      wrapper.instance().handleWidgetGoForward();

      expect(goForward).toHaveBeenCalledTimes(1);
      expect(goForward).toHaveBeenCalledWith();
    });

    it('when call handleWidgetRefresh', () => {
      wrapper.instance().handleWidgetRefresh();

      expect(reload).toHaveBeenCalledTimes(1);
      expect(reload).toHaveBeenCalledWith();
    });

    it('when call handleWidgetStopRefresh', () => {
      wrapper.instance().handleWidgetStopRefresh();

      expect(stop).toHaveBeenCalledTimes(1);
      expect(stop).toHaveBeenCalledWith();
    });
  });

  describe('test handleToggleSettingMenu', () => {
    const wrapper = mount(<WebWidget />);

    beforeEach(() => {
      wrapper.instance().webViewRef = webView;
    });

    it('when toggle state.isSettingOpen', () => {
      wrapper.instance().handleToggleSettingMenu();
      expect(wrapper.state().isSettingOpen).toBe(true);
      wrapper.instance().webViewRef = webView;
      wrapper.instance().handleToggleSettingMenu();
      expect(wrapper.state().isSettingOpen).toBe(false);
    });

    it('when set state.isSettingOpen = true ', () => {
      wrapper.instance().handleToggleSettingMenu(true);
      expect(wrapper.state().isSettingOpen).toBe(true);
    });

    it('when set state.isSettingOpen = false', () => {
      wrapper.instance().handleToggleSettingMenu(false);
      expect(wrapper.state().isSettingOpen).toBe(false);
    });
  });
});
