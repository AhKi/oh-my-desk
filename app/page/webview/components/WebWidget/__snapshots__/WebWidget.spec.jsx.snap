// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WebWidget /> should match to snapshot when render default 1`] = `
<WebWidget>
  <div
    className="WebWidget"
  >
    <WebWidgetHeader
      isLoading={false}
      isOnTop={false}
      onGoBack={[Function]}
      onGoForward={[Function]}
      onRefresh={[Function]}
      onStopRefresh={[Function]}
      onToggleIsOnTop={[Function]}
      onToggleSetting={[Function]}
      title=""
      webView={null}
    >
      <div
        className="WebWidgetHeader__title-bar"
      >
        <button
          className="WebWidgetHeader__thumbtack"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-thumbtack"
          />
        </button>
        <div
          className="WebWidgetHeader__history-set"
        >
          <HistoryGoBackButton
            className=""
            isCanGoBack={null}
            isCanGoForward={null}
            onGoBack={[Function]}
            onGoForward={[Function]}
          >
            <div
              className="HistoryGoBackButton"
            >
              <button
                className="HistoryGoBackButton__btn HistoryGoBackButton__btn--go-back"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fas fa-arrow-left"
                />
              </button>
              <button
                className="HistoryGoBackButton__btn HistoryGoBackButton__btn--go-front"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fas fa-arrow-right"
                />
              </button>
            </div>
          </HistoryGoBackButton>
          <ReloadButton
            className="ReloadButton"
            isLoading={false}
            onRefresh={[Function]}
            onStopRefresh={[Function]}
          >
            <div
              className="ReloadButton className"
            >
              <button
                className="ReloadButton__btn ReloadButton__btn--refresh"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fas fa-undo"
                />
              </button>
            </div>
          </ReloadButton>
        </div>
        <div
          className="WebWidgetHeader__title"
        />
        <div
          className="WebWidgetHeader__button-set"
        >
          <button
            className="WebWidgetHeader__button WebWidgetHeader__button-setting"
            onClick={[Function]}
            type="button"
          >
            <i
              className="fas fa-cog"
            />
          </button>
          <button
            className="WebWidgetHeader__button WebWidgetHeader__button--min"
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fas fa-window-minimize"
            />
          </button>
          <button
            className="WebWidgetHeader__button WebWidgetHeader__button--max"
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fas fa-square"
            />
          </button>
          <button
            className="WebWidgetHeader__button WebWidgetHeader__button--close"
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fas fa-times"
            />
          </button>
        </div>
      </div>
    </WebWidgetHeader>
    <WebWidgetMobileHeader
      isLoading={false}
      isOpen={true}
      onGoBack={[Function]}
      onGoForward={[Function]}
      onRefresh={[Function]}
      onStopRefresh={[Function]}
      onToggleSetting={[Function]}
      webView={null}
    >
      <div
        className="WebWidgetMobileHeader WebWidgetModalHeader__open"
      >
        <HistoryGoBackButton
          className=""
          isCanGoBack={null}
          isCanGoForward={null}
          onGoBack={[Function]}
          onGoForward={[Function]}
        >
          <div
            className="HistoryGoBackButton"
          >
            <button
              className="HistoryGoBackButton__btn HistoryGoBackButton__btn--go-back"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                className="fas fa-arrow-left"
              />
            </button>
            <button
              className="HistoryGoBackButton__btn HistoryGoBackButton__btn--go-front"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                className="fas fa-arrow-right"
              />
            </button>
          </div>
        </HistoryGoBackButton>
        <ReloadButton
          className="ReloadButton"
          isLoading={false}
          onRefresh={[Function]}
          onStopRefresh={[Function]}
        >
          <div
            className="ReloadButton className"
          >
            <button
              className="ReloadButton__btn ReloadButton__btn--refresh"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fas fa-undo"
              />
            </button>
          </div>
        </ReloadButton>
        <button
          className="WebWidgetMobileHeader__setting"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-cog"
          />
        </button>
      </div>
    </WebWidgetMobileHeader>
    <webview
      src="https://www.github.com"
      style={
        Object {
          "display": "inline-flex",
          "height": "100%",
          "overflow": "auto",
          "overflowY": "hidden",
          "width": "100%",
        }
      }
    />
  </div>
</WebWidget>
`;

exports[`<WebWidget /> should match to snapshot when state.isSettingOpen === true 1`] = `
<WebWidget>
  <div
    className="WebWidget"
  >
    <WebWidgetHeader
      isLoading={false}
      isOnTop={false}
      onGoBack={[Function]}
      onGoForward={[Function]}
      onRefresh={[Function]}
      onStopRefresh={[Function]}
      onToggleIsOnTop={[Function]}
      onToggleSetting={[Function]}
      title=""
      webView={
        Object {
          "canGoBack": [Function],
          "canGoForward": [Function],
          "goBack": [MockFunction],
          "goForward": [MockFunction],
          "reload": [MockFunction],
          "stop": [MockFunction],
        }
      }
    >
      <div
        className="WebWidgetHeader__title-bar"
      >
        <button
          className="WebWidgetHeader__thumbtack"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-thumbtack"
          />
        </button>
        <div
          className="WebWidgetHeader__history-set"
        >
          <HistoryGoBackButton
            className=""
            isCanGoBack={false}
            isCanGoForward={false}
            onGoBack={[Function]}
            onGoForward={[Function]}
          >
            <div
              className="HistoryGoBackButton"
            >
              <button
                className="HistoryGoBackButton__btn HistoryGoBackButton__btn--go-back"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fas fa-arrow-left"
                />
              </button>
              <button
                className="HistoryGoBackButton__btn HistoryGoBackButton__btn--go-front"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fas fa-arrow-right"
                />
              </button>
            </div>
          </HistoryGoBackButton>
          <ReloadButton
            className="ReloadButton"
            isLoading={false}
            onRefresh={[Function]}
            onStopRefresh={[Function]}
          >
            <div
              className="ReloadButton className"
            >
              <button
                className="ReloadButton__btn ReloadButton__btn--refresh"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fas fa-undo"
                />
              </button>
            </div>
          </ReloadButton>
        </div>
        <div
          className="WebWidgetHeader__title"
        />
        <div
          className="WebWidgetHeader__button-set"
        >
          <button
            className="WebWidgetHeader__button WebWidgetHeader__button-setting"
            onClick={[Function]}
            type="button"
          >
            <i
              className="fas fa-cog"
            />
          </button>
          <button
            className="WebWidgetHeader__button WebWidgetHeader__button--min"
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fas fa-window-minimize"
            />
          </button>
          <button
            className="WebWidgetHeader__button WebWidgetHeader__button--max"
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fas fa-square"
            />
          </button>
          <button
            className="WebWidgetHeader__button WebWidgetHeader__button--close"
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fas fa-times"
            />
          </button>
        </div>
      </div>
    </WebWidgetHeader>
    <WebWidgetMobileHeader
      isLoading={false}
      isOpen={true}
      onGoBack={[Function]}
      onGoForward={[Function]}
      onRefresh={[Function]}
      onStopRefresh={[Function]}
      onToggleSetting={[Function]}
      webView={
        Object {
          "canGoBack": [Function],
          "canGoForward": [Function],
          "goBack": [MockFunction],
          "goForward": [MockFunction],
          "reload": [MockFunction],
          "stop": [MockFunction],
        }
      }
    >
      <div
        className="WebWidgetMobileHeader WebWidgetModalHeader__open"
      >
        <HistoryGoBackButton
          className=""
          isCanGoBack={false}
          isCanGoForward={false}
          onGoBack={[Function]}
          onGoForward={[Function]}
        >
          <div
            className="HistoryGoBackButton"
          >
            <button
              className="HistoryGoBackButton__btn HistoryGoBackButton__btn--go-back"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                className="fas fa-arrow-left"
              />
            </button>
            <button
              className="HistoryGoBackButton__btn HistoryGoBackButton__btn--go-front"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                className="fas fa-arrow-right"
              />
            </button>
          </div>
        </HistoryGoBackButton>
        <ReloadButton
          className="ReloadButton"
          isLoading={false}
          onRefresh={[Function]}
          onStopRefresh={[Function]}
        >
          <div
            className="ReloadButton className"
          >
            <button
              className="ReloadButton__btn ReloadButton__btn--refresh"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fas fa-undo"
              />
            </button>
          </div>
        </ReloadButton>
        <button
          className="WebWidgetMobileHeader__setting"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-cog"
          />
        </button>
      </div>
    </WebWidgetMobileHeader>
    <WebWidgetSetting
      name=""
      onToggleSetting={[Function]}
      url=""
      widget={Object {}}
    >
      <div
        className="WebWidgetSetting"
      >
        <OutsideClickHandler
          onOutSideClick={[Function]}
        >
          <div
            className="OutsideClickHandler"
          >
            <form>
              <h3>
                Edit setting
              </h3>
              <ValidationInput
                autoFocus={true}
                className="ValidationInput"
                error=""
                errorClassName="ValidationInput__validation-message"
                inputClassName="ValidationInput__text-input"
                name="Name"
                nameClassName="ValidationInput__label"
                onChange={[Function]}
                placeholder=""
                value=""
              >
                <div
                  className="ValidationInput"
                >
                  <label
                    className="ValidationInput__label"
                    htmlFor="Name"
                  >
                    Name
                    <input
                      autoFocus={true}
                      className="ValidationInput__text-input"
                      id="Name"
                      onChange={[Function]}
                      placeholder=""
                      type="text"
                      value=""
                    />
                  </label>
                </div>
              </ValidationInput>
              <ValidationInput
                autoFocus={false}
                className="ValidationInput"
                error=""
                errorClassName="ValidationInput__validation-message"
                inputClassName="ValidationInput__text-input"
                name="Url"
                nameClassName="ValidationInput__label"
                onChange={[Function]}
                placeholder=""
                value=""
              >
                <div
                  className="ValidationInput"
                >
                  <label
                    className="ValidationInput__label"
                    htmlFor="Url"
                  >
                    Url
                    <input
                      autoFocus={false}
                      className="ValidationInput__text-input"
                      id="Url"
                      onChange={[Function]}
                      placeholder=""
                      type="text"
                      value=""
                    />
                  </label>
                </div>
              </ValidationInput>
              <button
                onClick={[Function]}
                type="button"
              >
                Go to manage view
              </button>
              <input
                onClick={[Function]}
                type="submit"
                value="Save"
              />
            </form>
          </div>
        </OutsideClickHandler>
      </div>
    </WebWidgetSetting>
    <webview
      src="https://www.github.com"
      style={
        Object {
          "display": "inline-flex",
          "height": "100%",
          "overflow": "auto",
          "overflowY": "hidden",
          "width": "100%",
        }
      }
    />
  </div>
</WebWidget>
`;
